## Table of contents

####  Step 1: Importing data file and libraries

#### Step 2: Consistency checks and data cleaning

#### Step 3: Descriptive statistics

## Step 1: Importing data file and libraries

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Importing data file by creating a path 

path = r'C:\Users\Lenovo\Documents\Eva\Achievement 6'
df = pd.read_csv(os.path.join(path, 'Project Achievement 6 Tornado.csv'), index_col = False)

In [3]:
# Checking the output

df.head()

Year  Month  Day        Date State  Magnitude  Injuries  Fatalities  \
0  1950      1    3  03.01.1950    IL          3         3           0   
1  1950      1    3  03.01.1950    MO          3         3           0   
2  1950      1    3  03.01.1950    OH          1         1           0   
3  1950      1   13  13.01.1950    AR          3         1           1   
4  1950      1   25  25.01.1950    IL          2         0           0   

   Starting latitude  Starting longitude  Ending latitude  Ending longitude  \
0              39.10              -89.30            39.12            -89.23   
1              38.77              -90.22            38.83            -90.03   
2              40.88              -84.58             0.00              0.00   
3              34.40              -94.37             0.00              0.00   
4              41.17              -87.33             0.00              0.00   

   Length of track  Width   
0              3.6     130  
1              9.5     150  
2              0.1      10  
3              0.6      17  
4              0.1     100

## Step 2: Consistency checks and data cleaning

In [4]:
# Numbers of rows and columns

df.shape

(67096, 14)

In [5]:
# Checking for mixed-type data columns

for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

No mixed-type data columns

In [6]:
# Finding missing values

df.isnull().sum()

Year                  0
Month                 0
Day                   0
Date                  0
State                 0
Magnitude             0
Injuries              0
Fatalities            0
Starting latitude     0
Starting longitude    0
Ending latitude       0
Ending longitude      0
Length of track       0
Width                 0
dtype: int64

No missing values found

In [7]:
# Finding missing values (value 0) in the column "Ending latitude"

df.loc[(df['Ending latitude']==0)]

Year  Month  Day        Date State  Magnitude  Injuries  Fatalities  \
2      1950      1    3  03.01.1950    OH          1         1           0   
3      1950      1   13  13.01.1950    AR          3         1           1   
4      1950      1   25  25.01.1950    IL          2         0           0   
11     1950      2   12  12.02.1950    AR          2         0           0   
13     1950      2   12  12.02.1950    LA          1         0           0   
...     ...    ...  ...         ...   ...        ...       ...         ...   
51560  2008     10    8  08.10.2008    AL          0         0           0   
51565  2008     10    9  09.10.2008    GA          0         0           0   
51566  2008     10   11  11.10.2008    KS          0         0           0   
51567  2008     10   11  11.10.2008    NM          0         0           0   
51568  2008     10   24  24.10.2008    GA          0         0           0   

       Starting latitude  Starting longitude  Ending latitude  \
2                  40.88              -84.58              0.0   
3                  34.40              -94.37              0.0   
4                  41.17              -87.33              0.0   
11                 34.48              -92.40              0.0   
13                 32.60              -91.33              0.0   
...                  ...                 ...              ...   
51560              33.93              -87.85              0.0   
51565              30.79              -82.56              0.0   
51566              39.33             -101.55              0.0   
51567              35.15             -105.94              0.0   
51568              31.02              -81.81              0.0   

       Ending longitude  Length of track  Width   
2                   0.0             0.10      10  
3                   0.0             0.60      17  
4                   0.0             0.10     100  
11                  0.0             0.10     100  
13                  0.0             0.50      33  
...                 ...              ...     ...  
51560               0.0             0.01      20  
51565               0.0             0.25      30  
51566               0.0             0.25      10  
51567               0.0             0.02      10  
51568               0.0             0.25      25  

[26023 rows x 14 columns]

There are 26023 missing values in the column "Ending latitude.

In [8]:
# Finding missing values (value 0) in the column "Ending latitude"

df.loc[(df['Ending longitude']==0)]

Year  Month  Day        Date State  Magnitude  Injuries  Fatalities  \
2      1950      1    3  03.01.1950    OH          1         1           0   
3      1950      1   13  13.01.1950    AR          3         1           1   
4      1950      1   25  25.01.1950    IL          2         0           0   
11     1950      2   12  12.02.1950    AR          2         0           0   
13     1950      2   12  12.02.1950    LA          1         0           0   
...     ...    ...  ...         ...   ...        ...       ...         ...   
51560  2008     10    8  08.10.2008    AL          0         0           0   
51565  2008     10    9  09.10.2008    GA          0         0           0   
51566  2008     10   11  11.10.2008    KS          0         0           0   
51567  2008     10   11  11.10.2008    NM          0         0           0   
51568  2008     10   24  24.10.2008    GA          0         0           0   

       Starting latitude  Starting longitude  Ending latitude  \
2                  40.88              -84.58              0.0   
3                  34.40              -94.37              0.0   
4                  41.17              -87.33              0.0   
11                 34.48              -92.40              0.0   
13                 32.60              -91.33              0.0   
...                  ...                 ...              ...   
51560              33.93              -87.85              0.0   
51565              30.79              -82.56              0.0   
51566              39.33             -101.55              0.0   
51567              35.15             -105.94              0.0   
51568              31.02              -81.81              0.0   

       Ending longitude  Length of track  Width   
2                   0.0             0.10      10  
3                   0.0             0.60      17  
4                   0.0             0.10     100  
11                  0.0             0.10     100  
13                  0.0             0.50      33  
...                 ...              ...     ...  
51560               0.0             0.01      20  
51565               0.0             0.25      30  
51566               0.0             0.25      10  
51567               0.0             0.02      10  
51568               0.0             0.25      25  

[26023 rows x 14 columns]

There are 26023 missing values in the column "Ending longitude.

In [9]:
# Finding missing values (value -9) in the column "Magnitude"

df.loc[(df['Magnitude']==-9)]

Year  Month  Day        Date State  Magnitude  Injuries  Fatalities  \
59886  2016      4   15  15.04.2016    CO         -9         0           0   
59887  2016      4   15  15.04.2016    CO         -9         0           0   
59888  2016      4   15  15.04.2016    CO         -9         0           0   
59889  2016      4   15  15.04.2016    CO         -9         0           0   
60023  2016      5    8  08.05.2016    OK         -9         0           0   
...     ...    ...  ...         ...   ...        ...       ...         ...   
66955  2021     12   15  15.12.2021    IA         -9         0           0   
66956  2021     12   15  15.12.2021    IA         -9         0           0   
66957  2021     12   15  15.12.2021    IA         -9         0           0   
66958  2021     12   15  15.12.2021    IA         -9         0           0   
66959  2021     12   15  15.12.2021    IA         -9         0           0   

       Starting latitude  Starting longitude  Ending latitude  \
59886            38.3743           -102.8185          38.4299   
59887            38.4046           -102.8083          38.4118   
59888            38.4773           -102.7234          38.4944   
59889            38.1773           -102.7907          38.1852   
60023            34.5810            -98.0760          34.5810   
...                  ...                 ...              ...   
66955            41.9551            -95.5749          42.0074   
66956            42.5620            -95.4050          42.5850   
66957            42.2525            -94.6317          42.2708   
66958            42.5797            -94.5775          42.5920   
66959            42.6852            -94.5062          42.7726   

       Ending longitude  Length of track  Width   
59886         -102.8225             3.85     150  
59887         -102.8082             0.50     100  
59888         -102.7191             1.20     100  
59889         -102.7865             0.59     100  
60023          -98.0530             1.30      40  
...                 ...              ...     ...  
66955          -95.5266             4.39      50  
66956          -95.3880             1.81      50  
66957          -94.6236             1.33      40  
66958          -94.5657             1.04      40  
66959          -94.4325             7.10      80  

[605 rows x 14 columns]

There are 605 missing values in the column "Magnitude".

In [18]:
# Replacing the missing values in the column "Magnitude" with "None"

df['Magnitude'].replace(-9, None, inplace = True)

In [19]:
# Finding missing values (value -9) in the column "Magnitude"

df.loc[(df['Magnitude']==-9)]

Empty DataFrame
Columns: [Year, Month, Day, Date, State, Magnitude, Injuries, Fatalities, Starting latitude, Starting longitude, Ending latitude, Ending longitude, Length of track, Width ]
Index: []

In [21]:
# Replacing the missing values in the columns "Ending latitude" with "None"

df['Ending latitude'].replace(0, None, inplace = True)

In [23]:
# Finding missing values (value 0) in the column "Ending latitude"

df.loc[(df['Ending latitude']==0)]

Empty DataFrame
Columns: [Year, Month, Day, Date, State, Magnitude, Injuries, Fatalities, Starting latitude, Starting longitude, Ending latitude, Ending longitude, Length of track, Width ]
Index: []

In [24]:
# Replacing the missing values in the columns "Ending longitude" with "None"

df['Ending longitude'].replace(0, None, inplace = True)

In [25]:
# Finding missing values (value 0) in the column "Ending longitude"

df.loc[(df['Ending longitude']==0)]

Empty DataFrame
Columns: [Year, Month, Day, Date, State, Magnitude, Injuries, Fatalities, Starting latitude, Starting longitude, Ending latitude, Ending longitude, Length of track, Width ]
Index: []

In [10]:
# Finding duplicate records

df_dups = df[df.duplicated()]

In [11]:
# Checking the output

df_dups

Empty DataFrame
Columns: [Year, Month, Day, Date, State, Magnitude, Injuries, Fatalities, Starting latitude, Starting longitude, Ending latitude, Ending longitude, Length of track, Width ]
Index: []

There are no duplicate records.

## Step 3: Descriptive statistics

In [33]:
df.dtypes

Year                    int64
Month                   int64
Day                     int64
Date                   object
State                  object
Magnitude              object
Injuries                int64
Fatalities              int64
Starting latitude     float64
Starting longitude    float64
Ending latitude        object
Ending longitude       object
Length of track       float64
Width                   int64
dtype: object

In [29]:
df.describe()

Year         Month           Day      Injuries    Fatalities  \
count  67096.000000  67096.000000  67096.000000  67096.000000  67096.000000   
mean    1991.427090      5.975751     15.928103      1.447344      0.091093   
std       19.323831      2.443085      8.739900     18.326231      1.489188   
min     1950.000000      1.000000      1.000000      0.000000      0.000000   
25%     1976.000000      4.000000      8.000000      0.000000      0.000000   
50%     1994.000000      6.000000     16.000000      0.000000      0.000000   
75%     2008.000000      7.000000     24.000000      0.000000      0.000000   
max     2021.000000     12.000000     31.000000   1740.000000    158.000000   

       Starting latitude  Starting longitude  Length of track        Width   
count       67096.000000        67096.000000     67096.000000  67096.000000  
mean           37.140897          -92.745491         3.490061    107.048319  
std             5.090188            8.687126         8.280171    206.300547  
min            17.721200         -163.530000         0.000000      0.000000  
25%            33.220000          -98.415550         0.110000     20.000000  
50%            37.020000          -93.530000         0.800000     50.000000  
75%            40.930000          -86.700000         3.180000    100.000000  
max            61.020000          -64.715100       234.700000   4576.000000

In [35]:
# Descriptive statistics (mode and median) for the other qualitative variables


df['Year'].median()

1994.0

In [36]:
df['Year'].mode()

0    2004
Name: Year, dtype: int64

In [37]:
df['Month'].median()

6.0

In [38]:
df['Month'].mode()

0    5
Name: Month, dtype: int64

In [39]:
df['Day'].median()

16.0

In [40]:
df['Day'].mode()

0    24
Name: Day, dtype: int64

In [42]:
df['State'].mode()

0    TX
Name: State, dtype: object

In [43]:
df['Date'].mode()

0    27.04.2011
Name: Date, dtype: object

In [45]:
df['Magnitude'].median()

1.0

In [46]:
df['Magnitude'].mode()

0    0
Name: Magnitude, dtype: object

In [47]:
# Exporting dataframe

df.to_csv(os.path.join(path, 'Project Achievement 6 Tornado_clean.csv'))